bundles/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/internal/handler/PentairControllerHandler.java

```diff
@@ -0,0 +1,1134 @@
+                    "Another controller is already configured.");
+        } else {
+            super.goOnline();
+        }
+    }
+
+    @Override
+    public void finishOnline() {
+        super.finishOnline();
+
+        // setup syncTimeJob to run once a day, initial time to sync is 3 minutes after controller goes online. This is
+        // to prevent collision with main thread queries on initial startup
+        syncTimeJob = scheduler.scheduleWithFixedDelay(() -> syncTime(), 3, 24 * 60 * 60, TimeUnit.MINUTES);
+
+        scheduler.execute(() -> readControllerSettings());
+    }
+
+    public void syncTime() {
+        boolean synctime = ((boolean) getConfig().get("synctime"));
+        if (synctime) {
+            logger.debug("Synchronizing System Time with Pentair controller");
+            Calendar now = Calendar.getInstance();
+            sendClockSettings(now.get(Calendar.HOUR_OF_DAY), now.get(Calendar.MINUTE), now.get(Calendar.DAY_OF_WEEK),
+                    now.get(Calendar.DAY_OF_MONTH), now.get(Calendar.MONTH) + 1, now.get(Calendar.YEAR) - 2000);
+        }
+    }
+
+    public void readControllerSettings() {
+        int i;
+
+        requestSWVersion();
+        requestHeatStatus();
+        requestClockSettings();
+
+        for (i = 1; i <= NUM_CIRCUITS; i++) {
+            requestCircuitNameFunction(i);
+        }
+
+        for (i = 1; i <= NUM_SCHEDULES; i++) {
+            requestSchedule(i);
+        }
+
+        requestLightGroups();
+        requestValves();
+    }
+
+    @Override
+    public void goOffline(ThingStatusDetail detail) {
+        super.goOffline(detail);
+
+        if (syncTimeJob != null) {
+            syncTimeJob.cancel(true);
+        }
+
+        onlineController = null;
+    }
+
+    public @Nullable PentairControllerCircuit getCircuitByGroupID(String groupID) {
+        for (ExpiringCache<PentairControllerCircuit> circuitItem : circuitsCache) {
+            PentairControllerCircuit circuit = circuitItem.getLastKnownValue();
+            if (circuit == null) {
+                break;
+            }
+
+            if (circuit.getGroupID().equals(groupID)) {
+                return circuit;
+            }
+        }
+
+        return null;
+    }
+
+    public int getScheduleNumber(String name) {
+        int i;
+
+        for (i = 1; i <= NUM_SCHEDULES; i++) {
+            String str = String.format(CONTROLLER_SCHEDULE, i);
+            if (str.equals(name)) {
+                return i;
+            }
+        }
+
+        return 0;
+    }
+
+    public Unit<Temperature> getUOM() {
+        return (this.uom) ? SIUnits.CELSIUS : ImperialUnits.FAHRENHEIT;
+    }
+
+    public QuantityType<Temperature> getWaterTemp() {
+        PentairControllerStatus status = this.controllerStatusCache.getLastKnownValue();
+        Objects.requireNonNull(status);
+        return new QuantityType<Temperature>(status.poolTemp, getUOM());
```

Comment:
Why throw an exception here if the status is null? Perhaps returning a UnDef status type would be better.